src/source.cpp

```diff
@@ -378,6 +410,96 @@ CompiledSourceWrapper::~CompiledSourceWrapper()
 #endif
 }
 
+//==============================================================================
+// MeshSource implementation
+//==============================================================================
+
+MeshSource::MeshSource(pugi::xml_node node)
+{
+  // Check for particle type
+  if (check_for_node(node, "particle")) {
+    auto temp_str = get_node_value(node, "particle", true, true);
+    if (temp_str == "neutron") {
+      particle_ = ParticleType::neutron;
+    } else if (temp_str == "photon") {
+      particle_ = ParticleType::photon;
+      settings::photon_transport = true;
+    } else {
+      fatal_error(std::string("Unknown source particle type: ") + temp_str);
+    }
+  }
+
+  int32_t mesh_id = stoi(get_node_value(node, "mesh"));
+  int32_t mesh_idx = model::mesh_map.at(mesh_id);
+  const auto& mesh = model::meshes[mesh_idx];
+
+  std::vector<double> strengths;
+  // read all source distributions and populate strengths vector for MeshSpatial
+  // object
+  for (auto source_node : node.children("source")) {
+    sources_.emplace_back(IndependentSource(source_node));
```

Comment:
This should be generic creation of a `Source` instead of `IndependentSource`

Reply:
Making a note here that I've changed the `MeshSource::sources_` vector to use `std::unique_ptr<Source>` objects to support the abstract `Source` class and the `Source::create` static method.